src/sidebar/util/fetch-config.js

```diff
@@ -132,9 +132,58 @@ async function fetchConfigRpc(appConfig, parentFrame, origin) {
     [],
     3000
   );
-  return fetchConfigEmbed(appConfig, remoteConfig);
+  const mergedConfig = fetchConfigEmbed(appConfig, remoteConfig);
+  return mutateAsyncGroups(mergedConfig, parentFrame, origin);
 }
 
+/**
+ * Potentially mutates the `groups` config object(s) to be a promise that
+ * resolves right away if the `groups` value exists in the original
+ * config, or returns a promise that resolves with a secondary RCP
+ * call to `requestGroups` when the `groups` value is '$rcp:requestGroups'
+ * If the `groups` value is missing or falsely then it won't be mutated.
+ *
+ * This allows the app to start with an incomplete config and then
+ * lazily fill in the `groups` value(s) later when its ready. This helps
+ * speed up the loading process.
+ *
+ * @param {Object} config - The configuration object to mutate. This should
+ *  already have the `services` value.
+ * @param {Window} parentFrame - Frame to send call to.
+ * @param {string} origin - Origin filter for `window.postMessage` call.
+ * @return {Object} - The mutated settings.
+ */
+
+async function mutateAsyncGroups(config, parentFrame, origin) {
+  if (Array.isArray(config.services)) {
+    config.services.forEach((service, index) => {
+      if (service.groups === '$rcp:requestGroups') {
+        // The `groups` need to be fetched from a secondary RCP call
+        service.groups = new Promise((resolve, reject) => {
+          postMessageJsonRpc
+            .call(
+              parentFrame,
+              origin,
+              'requestGroups',
+              [index], // Only supports one service for now
+              5000
+            )
+            .then(serviceGroupsRCP => {
+              resolve(serviceGroupsRCP);
+            })
+            .catch(() => {
+              reject('Unable to fetch groups');
+            });
+        });
+      } else if (service.groups) {
+        // The `groups` already exist so just wrap them in a promise so
+        // they can be treated identical do the RPC case above
+        service.groups = Promise.resolve(service.groups);
```

Comment:
Making `service.groups` always be a Promise for consistency seems reasonable. I will note that you can call `await` on a value, so you could actually leave `service.groups` as an array here and I think it would work just the same.

Reply:
awe right!